### PR TITLE
[XPU][DeepSeekV4] Deepseek-v4 inv-rope sycl kernel support

### PR DESCRIPTION
**Pull request by jjmiao1**
# Add DeepSeek-V4 Inverse RoPE SYCL Kernels

## Purpose

Add SYCL implementations for the DeepSeek-V4 **inverse RoPE** kernels used in the O-projection stage of the MLA pipeline. This PR includes two kernels:

1. **`deepseek_inv_rope_bf16`** — Inverse RoPE with bf16 output (for downstream bf16 einsum)
2. **`deepseek_inv_rope_fp8_quant`** — Inverse RoPE fused with FP8 quantization (for downstream fp8_einsum)

Both kernels perform:
- **NoPE dimensions (448):** Coalesced block copy using vectorized loads (`uint64_t`, 4 bf16 per lane)
- **RoPE dimensions (64):** Pair-wise inverse GPT-J rotation: `new_even = even * cos + odd * sin`, `new_odd = odd * cos - even * sin`
- **Layout transform:** Reshapes from `[T, H, D]` (per-head) to `[T, G, heads_per_group * D]` (group-contiguous)

---

### Part 1: `deepseek_inv_rope_bf16`

**Commit:** `8fecdeb` — Add Deepseek-v4 inv_rope bf16 SYCL support

#### Files Changed

| File | Change |
|------|--------|
| `csrc/xpu/sycl/deepseek_inv_rope_bf16.cpp` | New SYCL kernel (203 lines) |
| `csrc/xpu/ops.h` | Function declaration |
| `csrc/xpu/torch_bindings.cpp` | Op registration (`def` + `impl`) |
| `CMakeLists.txt` | Add source to build |
| `tests/test_deepseek_inv_rope_bf16.py` | Pytest with parametrized correctness + layout + edge case tests |

#### Design

- Grid: `(num_tokens, num_heads, SG_SIZE=16)`, one subgroup per head per token
- All model parameters (`heads_per_group`, `head_dim`, `nope_dim`, `rope_dim`) are runtime arguments
- Registered as `torch.ops._xpu_C.deepseek_inv_rope_bf16`

#### Unit Test — All PASSED

```bash
pytest tests/test_deepseek_inv_rope_bf16.py -v
```

- `test_correctness`: num_tokens ∈ {1,4,8,32,128,512} × num_heads/n_groups configs × seed ∈ {0,42}
- `test_output_layout`: validates group-contiguous output shape
- `test_zero_tokens`: empty-token edge case
- `test_positions_independence`: verifies position-wise correctness

#### Kernel Benchmark: SYCL vs Triton

Platform: Intel Arc Pro B60, Warmup 50 iters, Timed 200 iters

| num_tokens | io_bytes | triton(us) | sycl(us) | triton BW(GB/s) | sycl BW(GB/s) | speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 131,336 | 58.30 | 38.88 | 2.25 | 3.38 | **1.50x** |
| 2 | 262,672 | 57.44 | 38.10 | 4.57 | 6.89 | **1.51x** |
| 4 | 525,344 | 58.04 | 38.13 | 9.05 | 13.78 | **1.52x** |
| 8 | 1,050,688 | 57.02 | 38.13 | 18.43 | 27.55 | **1.50x** |
| 16 | 2,101,376 | 56.88 | 38.12 | 36.94 | 55.13 | **1.49x** |
| 32 | 4,202,752 | 57.04 | 38.14 | 73.69 | 110.18 | **1.50x** |
| 64 | 8,405,504 | 57.08 | 38.23 | 147.25 | 219.89 | **1.49x** |
| 128 | 16,811,008 | 56.75 | 38.63 | 296.22 | 435.21 | **1.47x** |
| 256 | 33,622,016 | 115.75 | 102.59 | 290.46 | 327.72 | **1.13x** |
| 512 | 67,244,032 | 215.06 | 193.04 | 312.67 | 348.34 | **1.11x** |
| 1024 | 134,488,064 | 443.57 | 361.97 | 303.20 | 371.55 | **1.23x** |
| 2048 | 268,976,128 | 885.83 | 702.28 | 303.64 | 383.00 | **1.26x** |
| 4096 | 537,952,256 | 1858.19 | 1369.68 | 289.50 | 392.76 | **1.36x** |
| 8192 | 1,075,904,512 | 4034.06 | 2740.93 | 266.70 | 392.53 | **1.47x** |
| 16384 | 2,151,809,024 | 8559.76 | 5395.05 | 251.39 | 398.85 | **1.59x** |

**SYCL achieves 1.13x–1.59x speedup.** Peak BW: 398.85 GB/s (87.5% of B60 HBM theoretical 456 GB/s).

#### Standalone SYCL Kernel (device time, event profiling)

Device: Intel Arc Pro B60 (HBM peak bandwidth: 456 GB/s)

| num_tokens | device(us) | BW(GB/s) | MBU |
|---:|---:|---:|---:|
| 1 | 1.90 | 69.06 | 15.1% |
| 2 | 1.88 | 139.73 | 30.6% |
| 4 | 1.94 | 270.49 | 59.3% |
| 8 | 2.34 | 448.41 | 98.3% |
| 16 | 2.52 | 834.59 | >100%* |
| 32 | 3.95 | 1063.59 | >100%* |
| 64 | 6.62 | 1270.29 | >100%* |
| 128 | 13.02 | 1290.88 | >100%* |
| 256 | 82.74 | 406.35 | 89.1% |
| 512 | 171.07 | 393.07 | 86.2% |
| 1024 | 340.02 | 395.53 | 86.7% |
| 2048 | 676.89 | 397.37 | 87.1% |
| 4096 | 1352.53 | 397.74 | 87.2% |
| 8192 | 2717.42 | 395.93 | 86.8% |
| 16384 | 5380.72 | 399.91 | 87.7% |

\* Small token counts show >100% MBU due to L2 cache hits (data fits in cache, not hitting HBM).

Large token regime stabilizes at **~87% MBU**, indicating near-optimal memory bandwidth utilization.

#### End-to-End Model Benchmark
**Note on E2E setup:** Due to current hardware resource constraints, we use a **10-layer trimmed checkpoint** of DeepSeek-V4-Flash-Base for E2E validation. This is sufficient to verify correct integration and no performance regression, as inv_rope executes identically in every decoder layer.

**Integration:** In `vllm/models/deepseek_v4/xpu/xpu_sparse.py` `_o_proj()`, the SYCL path calls `torch.ops._xpu_C.deepseek_inv_rope_bf16` to replace the original `rocm_inv_rope_einsum` baseline.

Model: DeepSeek-V4-Flash-Base (**10 layers**, trimmed), TP=4, input_len=2048, output_len=1024, max_concurrency=1, num_prompts=5, num_warmups=2

| Metric | baseline | SYCL | Delta |
|--------|------------------:|-----:|------:|
| Output token throughput (tok/s) | 8.11 | 8.41 | +3.7% |
| Mean TTFT (ms) | 2260.18 | 1902.40 | **-15.8%** |
| Median TTFT (ms) | 2250.30 | 1897.87 | **-15.7%** |
| P99 TTFT (ms) | 2293.55 | 1913.16 | **-16.6%** |
| Mean TPOT (ms) | 121.21 | 117.10 | **-3.4%** |
| Median TPOT (ms) | 121.42 | 116.83 | **-3.8%** |
| P99 TPOT (ms) | 121.61 | 117.74 | -3.2% |

SYCL inv_rope_bf16 shows **3.4% TPOT improvement** and **15.8% TTFT improvement** in E2E serving.

---

### Part 2: `deepseek_inv_rope_fp8_quant`

**Commit:** `52ad658` — Add Deepseek-v4 inv_rope fp8 quant SYCL support

#### Files Changed

| File | Change |
|------|--------|
| `csrc/xpu/sycl/deepseek_inv_rope_fp8_quant.cpp` | New SYCL kernel (272 lines) |
| `csrc/xpu/ops.h` | Function declaration |
| `csrc/xpu/torch_bindings.cpp` | Op registration (`def` + `impl`) |
| `CMakeLists.txt` | Add source to build |
| `tests/test_deepseek_inv_rope_fp8_quant.py` | Pytest with correctness + layout + FP8 range/scale tests |

#### Design

- Same grid/subgroup structure as bf16 variant
- Fuses inverse RoPE + per-group FP8 (E4M3) quantization in a single kernel pass
- Outputs: `fp8_tensor [T, G, heads_per_group * D]` + `scales [T, G]`
- Registered as `torch.ops._xpu_C.deepseek_inv_rope_fp8_quant`

#### Unit Test — All PASSED

```bash
pytest tests/test_deepseek_inv_rope_fp8_quant.py -v
```

- `test_correctness`: num_tokens ∈ {1,4,8,32,128} × num_heads/n_groups configs × seed ∈ {0,42}
- `test_output_layout`: validates FP8 output shape and scale shape
- `test_zero_tokens`: empty-token edge case
- `test_scales_are_power_of_two`: validates quantization scales
- `test_fp8_values_in_range`: validates FP8 output range [-448, 448]

#### Kernel Benchmark: SYCL vs Triton

Platform: Intel Arc Pro B60, Warmup 50 iters, Timed 200 iters

| num_tokens | io_bytes | triton(us) | sycl(us) | triton BW(GB/s) | sycl BW(GB/s) | speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 99,592 | 95.86 | 65.42 | 1.04 | 1.52 | **1.47x** |
| 2 | 199,184 | 96.80 | 65.44 | 2.06 | 3.04 | **1.48x** |
| 4 | 398,368 | 95.66 | 65.08 | 4.16 | 6.12 | **1.47x** |
| 8 | 796,736 | 95.20 | 65.38 | 8.37 | 12.19 | **1.46x** |
| 16 | 1,593,472 | 95.49 | 65.53 | 16.69 | 24.32 | **1.46x** |
| 32 | 3,186,944 | 95.66 | 65.39 | 33.32 | 48.74 | **1.46x** |
| 64 | 6,373,888 | 95.68 | 65.61 | 66.62 | 97.15 | **1.46x** |
| 128 | 12,747,776 | 102.01 | 65.43 | 124.96 | 194.84 | **1.56x** |
| 256 | 25,495,552 | 157.76 | 89.40 | 161.61 | 285.17 | **1.76x** |
| 512 | 50,991,104 | 256.75 | 168.34 | 198.60 | 302.90 | **1.53x** |
| 1024 | 101,982,208 | 446.25 | 303.55 | 228.53 | 335.97 | **1.47x** |
| 2048 | 203,964,416 | 824.53 | 579.29 | 247.37 | 352.09 | **1.42x** |
| 4096 | 407,928,832 | 1573.43 | 1118.10 | 259.26 | 364.84 | **1.41x** |
| 8192 | 815,857,664 | 3077.91 | 2212.07 | 265.07 | 368.82 | **1.39x** |
| 16384 | 1,631,715,328 | 6257.78 | 4310.08 | 260.75 | 378.58 | **1.45x** |

**SYCL achieves 1.39x–1.76x speedup.** Peak BW: 378.58 GB/s (83.0% of B60 HBM theoretical).

#### Standalone SYCL Kernel (device time, event profiling)

Device: Intel Arc Pro B60 (HBM peak bandwidth: 456 GB/s)

| num_tokens | device(us) | BW(GB/s) | MBU |
|---:|---:|---:|---:|
| 1 | 2.31 | 43.04 | 9.4% |
| 2 | 2.37 | 83.91 | 18.4% |
| 4 | 2.48 | 160.88 | 35.3% |
| 8 | 2.75 | 289.64 | 63.5% |
| 16 | 3.94 | 404.38 | 88.7% |
| 32 | 7.06 | 451.44 | 99.0% |
| 64 | 13.31 | 478.86 | >100%* |
| 128 | 24.91 | 511.76 | >100%* |
| 256 | 59.62 | 427.66 | 93.8% |
| 512 | 138.31 | 368.67 | 80.8% |
| 1024 | 273.54 | 372.82 | 81.8% |
| 2048 | 544.04 | 374.91 | 82.2% |
| 4096 | 1077.09 | 378.73 | 83.1% |
| 8192 | 2202.25 | 370.47 | 81.2% |
| 16384 | 4276.00 | 381.60 | 83.7% |

\* Small token counts show >100% MBU due to L2 cache hits (data fits in cache, not hitting HBM).

Large token regime stabilizes at **~81–84% MBU**.

#### End-to-End Model Benchmark

> The current model pipeline uses bf16 inv_rope only. FP8 quant variant will be integrated when the fp8 path is ready.

### CODE REVIEW

**Comment by jjmiao1** [comment]
@jikunshang please help review, thanks

File: csrc/xpu/sycl/deepseek_inv_rope_bf16.cpp [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+ *   2. Applies inverse GPT-J rotation on rope_dim elements using fp32
+ *      cos/sin cache.
+ *   3. Writes output in group-major layout: [T, G, hpg*D].
+ *
+ * Input:  o             [num_tokens, num_heads, head_dim] bf16
+ *         positions     [num_tokens] int64
+ *         cos_sin_cache [max_pos, rope_dim] float32  (cos||sin concatenated)
+ * Output: out           [num_tokens, n_groups, heads_per_group * head_dim] bf16
+ */
+class deepseek_inv_rope_bf16_kernel {
```
Comment (jikunshang) [inline csrc/xpu/sycl/deepseek_inv_rope_bf16.cpp:25]: there will never be a fp16 version, right?

File: csrc/xpu/ops.h [hunk 2]
```diff
@@ -92,6 +92,24 @@ void apply_rotary_emb(
     torch::Tensor& sin,     // [num_tokens, rot_dim/2]
     bool is_neox);
 
+torch::Tensor deepseek_inv_rope_bf16(
+    const torch::Tensor& o,
```
Comment (jikunshang) [inline csrc/xpu/ops.h]: what's `o` meaning here?

Comment (jjmiao1) [reply csrc/xpu/ops.h]: O is the attention output [num_tokens, num_heads, head_dim] bf16

Comment (jikunshang) [reply csrc/xpu/ops.h]: it's confusing, especially in kernel code, you have `o` and `fp8_out` `scale_out` 
please rename it.

Comment (jjmiao1) [reply csrc/xpu/ops.h]: Thanks for the comment, latest commit has renamed the `o` to `attn_output`.

File: csrc/xpu/sycl/deepseek_inv_rope_bf16.cpp (hunk 1 above)
Comment (jjmiao1) [reply csrc/xpu/sycl/deepseek_inv_rope_bf16.cpp:25]: Yes, no fp16 version is planned as there is no attention fp16 code path in vllm upstream DeepSeek-V4 implementation.

File: csrc/xpu/sycl/deepseek_inv_rope_fp8_quant.cpp [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+  using fp8_t = uint8_t;
+  constexpr int SG_SIZE = vllm::deepseek_inv_rope_fp8_quant_kernel::SG_SIZE;
+  constexpr int QUANT_GROUP_SIZE = vllm::QUANT_GROUP_SIZE;
+
+  TORCH_CHECK(o.dim() == 3, "o must be 3D [num_tokens, num_heads, head_dim]");
+  TORCH_CHECK(o.scalar_type() == torch::kBFloat16, "o must be bfloat16");
+  TORCH_CHECK(positions.scalar_type() == torch::kLong,
+              "positions must be int64");
+  TORCH_CHECK(cos_sin_cache.scalar_type() == torch::kFloat32,
+              "cos_sin_cache must be float32");
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply csrc/xpu/sycl/deepseek_inv_rope_fp8_quant.cpp]: Added  `TORCH_CHECK `validations. No need to do extra device check because this op is already registered on the XPU dispatch path, will not be called if the device is not XPU.

File: csrc/xpu/sycl/deepseek_inv_rope_fp8_quant.cpp [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+      {n_groups, num_tokens, num_scale_blocks},
+      o.options().dtype(torch::kFloat32));
+
+  if (num_tokens == 0) {
+    return {fp8_out, scale_out};
+  }
+
+  const int64_t o_stride_token = o.stride(0);
+  const int64_t o_stride_head = o.stride(1);
+  const int64_t cache_stride_pos = cos_sin_cache.stride(0);
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply csrc/xpu/sycl/deepseek_inv_rope_fp8_quant.cpp:271]: Added contiguous check.

File: csrc/xpu/sycl/deepseek_inv_rope_bf16.cpp [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    int64_t rope_dim) {
+  using bf16_t = sycl::ext::oneapi::bfloat16;
+  constexpr int SG_SIZE = vllm::deepseek_inv_rope_bf16_kernel::SG_SIZE;
+
+  TORCH_CHECK(o.dim() == 3, "o must be 3D [num_tokens, num_heads, head_dim]");
+  TORCH_CHECK(o.scalar_type() == torch::kBFloat16, "o must be bfloat16");
+  TORCH_CHECK(positions.scalar_type() == torch::kLong,
+              "positions must be int64");
+  TORCH_CHECK(cos_sin_cache.scalar_type() == torch::kFloat32,
+              "cos_sin_cache must be float32");
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply csrc/xpu/sycl/deepseek_inv_rope_bf16.cpp]: Added TORCH_CHECK validations.

File: csrc/xpu/sycl/deepseek_inv_rope_bf16.cpp [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+  auto out = torch::empty({num_tokens, n_groups, d},
+                          o.options().dtype(torch::kBFloat16));
+
+  if (num_tokens == 0) {
+    return out;
+  }
+
+  const int64_t o_stride_token = o.stride(0);
+  const int64_t o_stride_head = o.stride(1);
+  const int64_t cache_stride_pos = cos_sin_cache.stride(0);
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply csrc/xpu/sycl/deepseek_inv_rope_bf16.cpp:200]: Added contiguous check.

File: tests/test_deepseek_inv_rope_fp8_quant.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    # For power-of-2 floats, log2 should be exactly integer
+    log2_scales = torch.log2(scale_out)
+    is_integer = (log2_scales == log2_scales.round())
+    assert is_integer.all(), (
+        f"Not all scales are power-of-2. "
+        f"Non-pow2 count: {(~is_integer).sum().item()}")
+
+
+def test_fp8_values_in_range():
+    """Verify all FP8 values are within [-FP8_MAX, FP8_MAX] after dequant."""
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply tests/test_deepseek_inv_rope_fp8_quant.py]: fixed

File: tests/test_deepseek_inv_rope_bf16.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+Run:
+    pytest tests/test_deepseek_inv_rope_bf16.py -v
+"""
+import pytest
+import torch
+
+import vllm_xpu_kernels._xpu_C  # noqa: F401
+
+DEVICE = torch.device("xpu")
```
[the remark this answers is not in the document]
Comment (jjmiao1) [reply tests/test_deepseek_inv_rope_bf16.py:15]: No point add extra check since the tests will only run on XPU device.

**Review by Wei-Lin-Intel — APPROVED** [review]
LGTM

**Review by jikunshang — APPROVED** [review]

**Review by mayuyuace — APPROVED** [review]

**Comment by jikunshang** [comment]
pre commit fail again.
